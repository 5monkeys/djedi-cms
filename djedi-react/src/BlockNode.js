import React from "react";

export default function BlockNode() {
  return <span>BlockNode</span>;
}
