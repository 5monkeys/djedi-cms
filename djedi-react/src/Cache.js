export default class Cache {
  constructor({ ttl }) {
    this.ttl = ttl;
    this._cache = new Map();
  }

  get(key) {
    const item = this._cache.get(key);

    if (item == null) {
      return undefined;
    }

    const needsRefresh = Date.now() - item.timestamp > this.ttl;

    return { node: item.value, needsRefresh };
  }

  set(key, value) {
    this._cache.set(key, { timestamp: Date.now(), value });
  }

  delete(key) {
    this._cache.delete(key);
  }
}
