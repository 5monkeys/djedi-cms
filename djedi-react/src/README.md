<img alt="Djedi CMS" src="https://djedi-cms.org/_static/djedi-landscape.svg" width="500"/>

# djedi-react

[React] client for [Djedi CMS].

## Installation

```
npm install djedi-react
```

## Usage

Not yet.

## Development

* `npm start`: Start [Jest] in watch mode.
* `npm lint`: Run [ESLint] \(including [Prettier]).
* `npm lint:fix`: Autofix [ESLint] errors.
* `npm jest`: Run unit tests.
* `npm build`: Compile with [Babel].
* `npm test`: Check that everything works.
* `npm publish`: Publish to [npm], but only if `npm test` passes.

[babel]: http://babeljs.io/
[djedi cms]: http://djedi-cms.org/
[eslint]: https://eslint.org/
[jest]: https://facebook.github.io/
[npm]: https://www.npmjs.com/
[prettier]: https://prettier.io/
[react]: https://reactjs.org/
