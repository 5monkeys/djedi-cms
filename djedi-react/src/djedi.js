// This file does not contain anything React-specific. If need JS client for
// another framework this file could be extracted into its own package and be
// re-used.

import { applyUriDefaults, parseUri, stringifyUri } from "./uri";

/*
This class fetches and caches nodes, and keeps `window.DJEDI_NODES` up-todate.

Central interface:

    type Node = {
      // Input nodes: Possibly non-absolute URI.
      // Output nodes: Absolute, normalized URI.
      uri: string,
      // Input nodes: The default value, if any.
      // Output nodes: The final value, if any. The final value can be the
      // default value, or a new value entered by the user. In both cases the
      // backend might have rendered the value (such as markdown -> HTML).
      value: string | undefined,
    }

The admin sidebar expects the following mapping of all rendered nodes on the page:

    window.DJEDI_NODES = {
      "<uri>": "<default>",
    }
*/
export class Djedi {
  constructor() {
    this.options = makeDefaultOptions();

    this._nodes = new Map();
    this._renderedNodes = new Map();
    this._batch = makeEmptyBatch();
    this._DJEDI_NODES = {};

    if (typeof window !== "undefined") {
      if (window.DJEDI_NODES == null) {
        window.DJEDI_NODES = {};
      }
      this._DJEDI_NODES = window.DJEDI_NODES;
    }
  }

  get(passedNode, callback) {
    const node = this._normalizeNode(passedNode);
    const { uri } = node;
    const existingNode = this._nodes.get(uri);

    if (existingNode != null) {
      callback(existingNode);
      return;
    }

    this.loadMany([node]).then(
      results => {
        callback(results[uri] || new Error(`Missing result for node: ${uri}`));
      },
      error => {
        callback(error);
      }
    );
  }

  getBatched(passedNode, callback) {
    if (this.options.batchInterval <= 0) {
      this.get(passedNode, callback);
      return;
    }

    const node = this._normalizeNode(passedNode);
    const existingNode = this._nodes.get(node.uri);

    if (existingNode != null) {
      callback(existingNode);
      return;
    }

    const previous = this._batch.queue.get(node.uri) || {
      node,
      callbacks: [],
    };
    previous.callbacks.push(callback);

    if (this._batch.timeoutId != null) {
      return;
    }

    this._batch.timeoutId = setTimeout(
      this._flushBatch.bind(this),
      this.options.batchInterval
    );
  }

  async loadMany(nodes) {
    console.log("TODO: loadMany", nodes, this.options.baseUrl);
    try {
      const results = await Promise.resolve([]);
      this.addNodes(results);
      return results;
    } catch (passedError) {
      const error =
        passedError instanceof Error
          ? passedError
          : new Error(`Failed to load nodes: ${passedError}`);
      throw error;
    }
  }

  async loadByPrefix(prefixes) {
    const results = await Promise.resolve([]);
    console.log("TODO: preloadByPrefix", prefixes, results);
    this.addNodes(results);
    return results;
  }

  // Needed to pick up the results from `loadByPrefix` after server-side
  // rendering.
  addNodes(nodes) {
    for (const passedNode of nodes) {
      const node = this._normalizeNode(passedNode);
      this._nodes.set(node.uri, node);
    }
  }

  injectAdmin() {
    console.log("TODO: injectAdmin (maybe)");
  }

  reportRenderedNode(passedNode) {
    const node = this._normalizeNode(passedNode);
    const previous = this._renderedNodes.get(node.uri);

    if (previous != null) {
      if (previous.value !== node.value) {
        console.warn(
          "djedi-react: Rendering a node with with a different default value. That default will be ignored.",
          {
            uri: node.uri,
            prev: previous.value,
            next: node.value,
          }
        );
      }
      previous.numInstances++;
      return;
    }

    this._renderedNodes.set({
      value: node.value,
      numInstances: 1,
    });

    this._DJEDI_NODES[this._djediNodesUri(node.uri)] = node.value;
  }

  reportRemovedNode(passedUri) {
    const uri = this._normalizeUri(passedUri);
    const previous = this._renderedNodes.get(uri);

    if (previous == null) {
      return;
    }

    previous.numInstances--;

    if (previous.numInstances <= 0) {
      this._renderedNodes.delete(uri);
      delete this._DJEDI_NODES[this._djediNodesUri(uri)];
    }
  }

  resetOptions() {
    this.options = makeDefaultOptions();
  }

  resetNodes() {
    if (this._batch.timeoutId != null) {
      clearTimeout(this._batch.timeoutId);
    }

    this._nodes = new Map();
    this._renderedNodes = new Map();
    this._batch = makeEmptyBatch();
    this._DJEDI_NODES = {};

    if (typeof window !== "undefined") {
      window.DJEDI_NODES = this._DJEDI_NODES;
    }
  }

  element(uri) {
    const uriObject = this._parseUri(uri);
    return {
      tag: "span",
      attributes: {
        "data-i18n": this._stringifyUri({
          ...uriObject,
          scheme: "",
          ext: "",
          version: "",
        }),
      },
    };
  }

  _parseUri(uri) {
    const { defaults, namespaceByScheme, separators } = this.options.uri;
    const uriObject = parseUri(uri, separators);
    return applyUriDefaults(uriObject, defaults, namespaceByScheme);
  }

  _stringifyUri(uriObject) {
    return stringifyUri(uriObject, this.options.uri.separators);
  }

  _normalizeUri(uri) {
    return this._stringifyUri(this._parseUri(uri));
  }

  _normalizeNode(node) {
    return {
      ...node,
      uri: this._stringifyUri(this._parseUri(node.uri)),
    };
  }

  _djediNodesUri(uri) {
    const uriObject = this._parseUri(uri);
    return this._stringifyUri({ ...uriObject, version: "" });
  }

  async _flushBatch() {
    const { queue } = this._batch;
    const nodes = Array.from(queue).reduce((result, [uri, data]) => {
      result[uri] = data.node.value;
      return result;
    });

    this._batch = makeEmptyBatch();

    try {
      const results = this.loadMany(nodes);
      for (const [uri, data] of queue) {
        const value = results[uri];
        const node =
          value == null
            ? new Error(`Missing result for node: ${uri}`)
            : { uri, value };
        for (const callback of data.callbacks) {
          callback(node);
        }
      }
    } catch (error) {
      for (const [, data] of queue) {
        for (const callback of data.callbacks) {
          callback(error);
        }
      }
    }
  }
}

// This is a function, not a constant, since it can be mutated by the user.
function makeDefaultOptions() {
  return {
    baseUrl: "",
    batchInterval: 50, // ms
    defaultRender: {
      loading: "Loadingâ€¦",
      error: "Failed to fetch content ðŸ˜ž",
    },
    uri: {
      defaults: {
        scheme: "i18n",
        namespace: "",
        path: "",
        ext: "txt",
        version: "",
      },
      namespaceByScheme: {
        i18n: "en-us",
        l10n: "local",
        g11n: "global",
      },
      separators: {
        scheme: "://",
        namespace: "@",
        path: "/",
        ext: ".",
        version: "#",
      },
    },
  };
}

// This is a function, not a constant, since it will be mutated.
function makeEmptyBatch() {
  return {
    timeoutId: undefined,
    queue: new Map(),
  };
}

export default new Djedi();
