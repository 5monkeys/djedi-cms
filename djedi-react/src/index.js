export { default as djedi } from "./djedi";
export { default as Node, NodeContext } from "./Node";
export { default as md } from "dedent-js";
