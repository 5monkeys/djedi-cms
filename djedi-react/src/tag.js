export { default } from "dedent-js";
