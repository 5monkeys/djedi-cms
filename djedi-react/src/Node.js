import PropTypes from "prop-types";
import React from "react";
import dedent from "dedent";

import djedi from "./djedi";

const POTENTIAL_HTML = /[<&]/;

const propTypes = {
  uri: PropTypes.string.isRequired,
  children: PropTypes.string, // Yes, a *string*!
  edit: PropTypes.bool,
  render: PropTypes.func,
  // ...kwargs: {[key: string]: string}.
};

const defaultProps = {
  edit: true,
  children: undefined,
  render: undefined,
};

export default class Node extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      node: undefined,
    };

    this.mounted = false;

    // Must be done in the constructor rather than `componentDidMount` because
    // of server-side rendering.
    this._get();
  }

  componentDidMount() {
    this.mounted = true;
    djedi.reportRenderedNode({
      uri: this.props.uri,
      value: this._getDefault(),
    });
  }

  componentDidUpdate(prevProps) {
    if (this.props.uri !== prevProps.uri) {
      this._get();
      djedi.reportRemovedNode(prevProps.uri);
      djedi.reportRenderedNode({
        uri: this.props.uri,
        value: this._getDefault(),
      });
    }

    if (
      this.props.uri === prevProps.uri &&
      this.props.children !== prevProps.children
    ) {
      console.error(
        "djedi-react: Changing the default value of a node is not supported.",
        { prev: prevProps.children, next: this.props.children }
      );
    }
  }

  componentWillUnmount() {
    this.mounted = false;
    djedi.reportRemovedNode(this.props.uri);
  }

  _get() {
    djedi.getBatched(
      { uri: this.props.uri, value: this._getDefault() },
      node => {
        if (this.mounted) {
          this.setState({ node });
        } else {
          // If the node already exists in cache, this callback is called
          // immediately (synchronously). If this happens in the `_get()` call in
          // `constructor`, where it is not allowed to call `setState`, mutate
          // `this.state` directly instead (which is fine in the `constructor`).
          // eslint-disable-next-line react/no-direct-mutation-state
          this.state.node = node;
        }
      }
    );
  }

  _getDefault() {
    const { children } = this.props;
    return children == null ? undefined : dedent(String(children));
  }

  render() {
    const {
      uri,
      children,
      edit,
      // Using a destructuring default rather than `defaultProps` for `render`,
      // since `djedi.options` may change.
      render = djedi.options.defaultRender,
      // Make sure to destructure all props above (even ones unused in this
      // method) so that `kwargs` only contains non-props.
      ...kwargs
    } = this.props;
    const { node } = this.state;

    if (node == null) {
      return render({ type: "loading" });
    }

    if (node instanceof Error) {
      return render({ type: "error", error: node });
    }

    const value = interpolate(node.value || "", kwargs);
    const element = djedi.element(node.uri);

    // This is mostly to make the test snapshots easier to read. Might be faster
    // not using `dangerouslySetInnerHTML` unnecessarily, too.
    const hasHtml = POTENTIAL_HTML.test(value);

    const content = edit ? (
      <element.tag
        {...element.attributes}
        dangerouslySetInnerHTML={hasHtml ? { __html: value } : undefined}
      >
        {hasHtml ? undefined : value}
      </element.tag>
    ) : (
      // WARNING: It’s not possible to render HTML without a wrapper element in
      // React. So you can’t do `edit={false}` and expect the node value to be
      // treated as HTML.
      value
    );

    return render({ type: "success", content });
  }
}

Node.propTypes = propTypes;
Node.defaultProps = defaultProps;

const INNER = /[^{}[\]\s]+/.source;
const INTERPOLATION_REGEX = RegExp(`\\{${INNER}\\}|\\[${INNER}\\]`, "g");

/*
This intentionally only supports `{key}`, not any of the fancy Python string
formatting extras. If `key` maps to a string in `kwargs`, that string is
inserted. Otherwise it is kept as-is. No escaping, no nothing. KISS and
YAGNI. If somebody does want fancy extras, they can do it themselves before
putting the value in `kwargs`.

Also support `[key]`, since `{foo}` is already used in JSX syntax:

    const user = "Bob"
    <Node uri="test">Hello, {user}!</Node>
    <Node uri="test">Hello, [user]!</Node>
    <Node uri="test">{`Hello, {user}!`}</Node>
*/
function interpolate(string, kwargs) {
  return string.replace(INTERPOLATION_REGEX, match => {
    const key = match.slice(1, -1);
    return Object.prototype.hasOwnProperty.call(kwargs, key)
      ? kwargs[key]
      : match;
  });
}
