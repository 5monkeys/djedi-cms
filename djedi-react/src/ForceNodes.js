import PropTypes from "prop-types";
import React from "react";

import Node from "./Node";

ForceNodes.propTypes = {
  children: PropTypes.node.isRequired,
};

export default function ForceNodes({ children }) {
  return React.Children.map(children, child => {
    if (child == null || child.type !== Node) {
      throw new TypeError(
        "djedi-react: <ForceNodes> only accepts <Node>s as children."
      );
    }
    return React.cloneElement(child, { render: () => null });
  });
}
