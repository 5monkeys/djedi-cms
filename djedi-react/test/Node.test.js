import React from "react";
import renderer from "react-test-renderer";

import { Node, djedi } from "../src";

import { fetch, resetAll, wait } from "./helpers";

jest.useFakeTimers();

beforeEach(resetAll);

test("it renders loading and then the node", async () => {
  fetch({ "i18n://en-us@test.txt": "returned text" });
  const component = renderer.create(<Node uri="test" />);
  expect(component.toJSON()).toMatchSnapshot("loading");
  await wait();
  expect(fetch.mockFn.mock.calls).toMatchSnapshot("api call");
  expect(component.toJSON()).toMatchSnapshot("with value");
  expect(window.DJEDI_NODES).toMatchSnapshot("window.DJEDI_NODES");
});

test("it renders synchronously if the node is already in cache", async () => {
  fetch({ "i18n://en-us@home/intro.txt": "Welcome to our amazing website!" });
  const nodes = await djedi.loadByPrefix(["home/"]);
  expect(fetch.mockFn.mock.calls).toMatchSnapshot("api call");
  expect(nodes).toMatchSnapshot("loadByPrefix result");
  const component = renderer.create(
    <Node uri="home/intro">Hello, World!</Node>
  );
  expect(component.toJSON()).toMatchSnapshot("with value");
});
