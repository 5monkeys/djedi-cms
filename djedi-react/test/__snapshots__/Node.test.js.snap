// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`batching: api calls (one request for 1 & 2, one for 3) 1`] = `
Array [
  Array [
    "/djedi/load_many",
    Object {
      "body": Object {
        "i18n://en-us@1.txt": null,
        "i18n://en-us@2.txt": null,
      },
      "headers": Object {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  Array [
    "/djedi/load_many",
    Object {
      "body": Object {
        "i18n://en-us@3.txt": null,
      },
      "headers": Object {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`custom render function: api calls 1`] = `
Array [
  Array [
    "/djedi/load_many",
    Object {
      "body": Object {
        "i18n://en-us@1.txt": null,
      },
      "headers": Object {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  Array [
    "/djedi/load_many",
    Object {
      "body": Object {
        "i18n://en-us@2.txt": null,
      },
      "headers": Object {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  Array [
    "/djedi/load_many",
    Object {
      "body": Object {
        "i18n://en-us@3.txt": null,
      },
      "headers": Object {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  Array [
    "/djedi/load_many",
    Object {
      "body": Object {
        "i18n://en-us@4.txt": null,
      },
      "headers": Object {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`custom render function: loading 1`] = `
<div>
  loading
  <hr />
  LOADING
</div>
`;

exports[`custom render function: missing 1`] = `
<div>
  error
  <hr />
  <div
    data-details={
      Object {
        "message": "Missing result for node: i18n://en-us@3.txt",
        "responseText": undefined,
        "status": 1404,
      }
    }
  >
    Error
  </div>
</div>
`;

exports[`custom render function: network error 1`] = `
<div>
  error
  <hr />
  <div
    data-details={
      Object {
        "message": "Djedi API error for request: POST /djedi/load_many
RequestData sent: {
  \\"i18n://en-us@2.txt\\": null
}
Response: undefined
Error: Network error",
        "responseText": "",
        "status": -1,
      }
    }
  >
    Error
  </div>
</div>
`;

exports[`custom render function: status error 1`] = `
<div>
  error
  <hr />
  <div
    data-details={
      Object {
        "message": "Djedi API error for request: POST /djedi/load_many
RequestData sent: {
  \\"i18n://en-us@1.txt\\": null
}
Response: 500 <mock.statusText>
Error: Unexpected response status code. Got 500 but expected 200 <= status < 400.",
        "responseText": "<h1>Server error 500</h1>",
        "status": 500,
      }
    }
  >
    Error
  </div>
</div>
`;

exports[`custom render function: with value 1`] = `
<div>
  success
  <hr />
  <article>
    <span
      data-i18n="en-us@4"
    >
      returned value
    </span>
  </article>
</div>
`;

exports[`default values are dedented: api call 1`] = `
Array [
  Array [
    "/djedi/load_many",
    Object {
      "body": Object {
        "i18n://en-us@first.txt": "Lorem ipsum dolor sit amet consectetur adipiscing elit tempor id, hendrerit euismod iaculis convallis ante tincidunt tempus bibendum metus, torquent ac egestas integer erat pharetra vehicula senectus.",
        "i18n://en-us@second.txt": "Lorem ipsum dolor sit amet consectetur adipiscing elit tempor id,
hendrerit euismod iaculis convallis ante tincidunt tempus bibendum
metus, torquent ac egestas integer erat pharetra vehicula senectus.",
      },
      "headers": Object {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`default values are dedented: with value 1`] = `
<div>
  <span
    data-i18n="en-us@first"
  >
    first
  </span>
  <span
    data-i18n="en-us@second"
  >
    second
  </span>
</div>
`;

exports[`edge case: if node.value is missing somehow, it doesnâ€™t crash 1`] = `
<span
  data-i18n="en-us@test"
>
  
</span>
`;

exports[`edit=false: with value 1`] = `"test"`;

exports[`interpolations: passing values to backend: api call 1`] = `
Array [
  Array [
    "/djedi/load_many",
    Object {
      "body": Object {
        "i18n://en-us@1.txt": "Hello, ,Bob,!",
        "i18n://en-us@2.txt": "Hello, {user}!",
        "i18n://en-us@3.txt": "Hello, [user]!",
        "i18n://en-us@4.txt": "Hello, {user}!",
        "i18n://en-us@5.txt": "Hello, [user]!",
        "i18n://en-us@6.txt": "Hello,
{user}!",
        "i18n://en-us@7.txt": "Hello,
[user]!",
      },
      "headers": Object {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`interpolations: passing values to backend: console.error (for uri \`1\`) 1`] = `
Array [
  Array [
    "Warning: Failed prop type: Invalid prop \`children\` of type \`array\` supplied to \`NodeInterpolationMistake\`, expected \`string\`.
    in NodeInterpolationMistake",
  ],
]
`;

exports[`interpolations: rendering values: rendered 1`] = `
<span
  dangerouslySetInnerHTML={
    Object {
      "__html": "Hello Bob!
Hello Bob!
Number: 3.141592653589793
Number: 3.141592653589793
Formatted number: 3.14
Formatted number: 3.14
Date: Wed Jun 27 2018 02:00:00 GMT+0200 (CEST)
Date: Wed Jun 27 2018 02:00:00 GMT+0200 (CEST)
hasOwnProperty: Asking for trouble
hasOwnProperty: Asking for trouble
undefined: undefined
undefined: undefined
null: null
null: null
Unusual key: slash
Unusual key: slash
Whitespace in key not interpolated: {a b}
Whitespace in key not interpolated: [a b]
Delimiter in key not interpolated: {a{b}
Delimiter in key not interpolated: {a}b}
Delimiter in key not interpolated: [a[b]
Delimiter in key not interpolated: [a]b]
Empty key not interpolated: {}
Empty key not interpolated: []
edit not interpolated: {edit}
edit not interpolated: [edit]
children not interpolated: {children}
children not interpolated: [children]
toString not interpolated: {toString}
toString not interpolated: [toString]
Unknown key left alone: {unknown}
Unknown key left alone: [unknown]
Escaped: &#x7b;name&#x7d;
Escaped: &#x5b;name&#x5d;
Double brackets OK: {Bob}
Double brackets OK: [Bob]
Double brackets OK: [Bob]
Double brackets OK: {Bob}",
    }
  }
  data-i18n="en-us@test"
/>
`;

exports[`it allows changing the default/children if also changing the uri: api call 1`] = `
Array [
  Array [
    "/djedi/load_many",
    Object {
      "body": Object {
        "i18n://en-us@first.txt": "first",
      },
      "headers": Object {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  Array [
    "/djedi/load_many",
    Object {
      "body": Object {
        "i18n://en-us@second.txt": "second",
      },
      "headers": Object {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`it allows changing the default/children if also changing the uri: render 1`] = `
<span
  data-i18n="en-us@first"
>
  first
</span>
`;

exports[`it allows changing the default/children if also changing the uri: render 2`] = `
<span
  data-i18n="en-us@second"
>
  second
</span>
`;

exports[`it fetches again after changing the uri prop (but not other props): api call 1`] = `
Array [
  Array [
    "/djedi/load_many",
    Object {
      "body": Object {
        "i18n://en-us@first.txt": null,
      },
      "headers": Object {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  Array [
    "/djedi/load_many",
    Object {
      "body": Object {
        "i18n://en-us@second.txt": null,
      },
      "headers": Object {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`it fetches again after changing the uri prop (but not other props): first render 1`] = `
<span
  data-i18n="en-us@first"
>
  first
</span>
`;

exports[`it fetches again after changing the uri prop (but not other props): second render 1`] = `
<span
  data-i18n="en-us@second"
>
  second
</span>
`;

exports[`it fetches again after changing the uri prop (but not other props): third render (same uri, no edit) 1`] = `"second"`;

exports[`it handles error status codes: error 1`] = `"Failed to fetch content ðŸ˜ž (500)"`;

exports[`it handles missing nodes in response: missing 1`] = `"Failed to fetch content ðŸ˜ž (1404)"`;

exports[`it handles rejected requests: error 1`] = `"Failed to fetch content ðŸ˜ž (-1)"`;

exports[`it handles window.DJEDI_NODES: before render 1`] = `Object {}`;

exports[`it handles window.DJEDI_NODES: render 1`] = `
Object {
  "i18n://en-us@changing1.txt": undefined,
  "i18n://en-us@loneRemoved.txt": "loneRemoved",
  "i18n://en-us@removed.txt": undefined,
}
`;

exports[`it handles window.DJEDI_NODES: render 2`] = `
Object {
  "i18n://en-us@changing1.txt": undefined,
  "i18n://en-us@changing2.txt": undefined,
  "i18n://en-us@removed.txt": undefined,
}
`;

exports[`it handles window.DJEDI_NODES: render 3`] = `
Object {
  "i18n://en-us@changing1.txt": undefined,
  "i18n://en-us@changing3.txt": undefined,
  "i18n://en-us@removed.txt": undefined,
}
`;

exports[`it handles window.DJEDI_NODES: unmounted 1`] = `Object {}`;

exports[`it renders loading and then the node: api call 1`] = `
Array [
  Array [
    "/djedi/load_many",
    Object {
      "body": Object {
        "i18n://en-us@test.txt": null,
      },
      "headers": Object {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`it renders loading and then the node: loading 1`] = `"Loadingâ€¦"`;

exports[`it renders loading and then the node: window.DJEDI_NODES 1`] = `
Object {
  "i18n://en-us@test.txt": undefined,
}
`;

exports[`it renders loading and then the node: with value 1`] = `
<span
  data-i18n="en-us@test"
>
  returned text
</span>
`;

exports[`it renders synchronously if the node is already in cache: api call 1`] = `
Array [
  Array [
    "/djedi/load_by_prefix",
    Object {
      "body": Array [
        "home/",
      ],
      "headers": Object {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`it renders synchronously if the node is already in cache: browser 1`] = `
<span
  data-i18n="en-us@home/intro"
>
  Welcome to our amazing website!
</span>
`;

exports[`it renders synchronously if the node is already in cache: loadByPrefix result 1`] = `
Object {
  "i18n://en-us@home/intro.txt": "Welcome to our amazing website!",
}
`;

exports[`it renders synchronously if the node is already in cache: with value 1`] = `
<span
  data-i18n="en-us@home/intro"
>
  Welcome to our amazing website!
</span>
`;

exports[`it treats node values as HTML: with value 1`] = `
<div>
  <span
    dangerouslySetInnerHTML={
      Object {
        "__html": "A <a href=\\"http://example.com\\">link</a>.",
      }
    }
    data-i18n="en-us@first"
  />
  <span
    dangerouslySetInnerHTML={
      Object {
        "__html": "A &amp; B",
      }
    }
    data-i18n="en-us@second"
  />
</div>
`;

exports[`it warns about rendering nodes with different defaults: api call 1`] = `
Array [
  Array [
    "/djedi/load_many",
    Object {
      "body": Object {
        "i18n://en-us@test.txt": "default",
      },
      "headers": Object {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`it warns about rendering nodes with different defaults: console.warn 1`] = `
Array [
  Array [
    "djedi-react: Rendering a node with with a different default value. That default will be ignored.",
    Object {
      "next": "other default",
      "prev": "default",
      "uri": "i18n://en-us@test.txt",
    },
  ],
  Array [
    "djedi-react: Rendering a node with with a different default value. That default will be ignored.",
    Object {
      "next": undefined,
      "prev": "default",
      "uri": "i18n://en-us@test.txt",
    },
  ],
]
`;

exports[`it warns about rendering nodes with different defaults: window.DJEDI_NODES 1`] = `
Object {
  "i18n://en-us@test.txt": "default",
}
`;

exports[`it warns if changing the default/children: api call 1`] = `
Array [
  Array [
    "/djedi/load_many",
    Object {
      "body": Object {
        "i18n://en-us@test.txt": "first",
      },
      "headers": Object {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`it warns if changing the default/children: console.error 1`] = `
Array [
  Array [
    "djedi-react: Changing the default value of a node is not supported.",
    Object {
      "next": "second",
      "prev": "first",
    },
  ],
]
`;

exports[`it warns if changing the default/children: first render 1`] = `
<span
  data-i18n="en-us@test"
>
  test value
</span>
`;

exports[`it warns if changing the default/children: second render (same as first render) 1`] = `
<span
  data-i18n="en-us@test"
>
  test value
</span>
`;

exports[`it warns when passing a non-string as default/children: api call 1`] = `
Array [
  Array [
    "/djedi/load_many",
    Object {
      "body": Object {
        "i18n://en-us@test.txt": "A ,[object Object]",
      },
      "headers": Object {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`it warns when passing a non-string as default/children: console.error 1`] = `
Array [
  Array [
    "Warning: Failed prop type: Invalid prop \`children\` of type \`array\` supplied to \`NodeChildrenMistake\`, expected \`string\`.
    in NodeChildrenMistake",
  ],
]
`;

exports[`it warns when passing a non-string as default/children: loading 1`] = `"Loadingâ€¦"`;

exports[`it warns when passing a non-string as default/children: with value 1`] = `
<span
  data-i18n="en-us@test"
>
  returned text
</span>
`;

exports[`the md tag ignores interpolations and warns about them: api call 1`] = `
Array [
  Array [
    "/djedi/load_many",
    Object {
      "body": Object {
        "i18n://en-us@test.txt": "Hello, !",
      },
      "headers": Object {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`the md tag ignores interpolations and warns about them: console.warn 1`] = `
Array [
  Array [
    "djedi-react: Using \`\${foo}\` in a node default value is an anti-pattern: it won't work if the user edits the node. For consistency, your interpolations will be ignored. Did you mean to use \`{foo}\` (without the \`$\`) or \`[foo]\`?",
  ],
]
`;

exports[`the md tag ignores interpolations and warns about them: with value 1`] = `
<span
  data-i18n="en-us@test"
>
  test
</span>
`;

exports[`using the md tag: api call 1`] = `
Array [
  Array [
    "/djedi/load_many",
    Object {
      "body": Object {
        "i18n://en-us@test.txt": "# A heading

Some text with a [link](http://example.com)

> Blockquote

    function codeBlock() {
      return \\"with indentation\\"
    }",
      },
      "headers": Object {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`using the md tag: with value 1`] = `
<span
  data-i18n="en-us@test"
>
  test
</span>
`;
